File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    answer = 0
    graph = [[] for _ in range(n)]
    visited = [False]*n
    
    for i in range(n):
        for j in range(n):
            if i != j and computers[i][j] == 1:
                graph[i].append(j)
    
    def bfs(x):
        q = deque([x])
        
        while q:
            now = q.popleft()
            visited[now] = True
            
            for X in graph[now]:
                if not visited[X]:
                    q.append(X)
                    visited[X] = True
               
    for i in range(n):
        if not visited[i]:
            bfs(i)
            answer+=1
        
    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;
class Solution {
    static ArrayList<Integer>[] edge;
    static boolean[] visited;
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        visited = new boolean[n];
        edge = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            edge[i] = new ArrayList<>();
        }
        
        for (int i=0;i<n;i++) {
            for (int j=0;j<n;j++) {
                if (i==j) continue;
                if (computers[i][j] == 1) {
                    edge[i].add(j);
                    edge[j].add(i);
                }
            }
        }
        
        for (int i=0;i<n;i++) {
            if (!visited[i]) {
                bfs(i);
                answer++;
            }
        }
        return answer;
    }
    
    private static void bfs(int now) {
        Queue<Integer> q = new ArrayDeque<>();
        visited[now] = true;
        q.add(now);
        
        while (!q.isEmpty()) {
            int x = q.poll();
            for (int nxt: edge[x]) {
                if (!visited[nxt]) {
                    visited[nxt]=true;
                    q.add(nxt);
                }
            }
        }
    }
}